test: avoid test-cluster-master-kill flakiness

stefanmb opened the pull request:
<!--
Thank you for submitting a pull request to Node.js. Before you submit, please
review below requirements and walk through the checklist. You can 'tick'
a box by using the letter "x": [x].

Run the test suite by invoking: `make -j4 lint test` on linux or
`vcbuild test nosign` on Windows.

If this aims to fix a regression or you’re adding a feature, make sure you also
write a test. Finally – if possible – a benchmark that quantifies your changes.

Finally, read through our contributors guide and make adjustments as necessary:
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist

<!-- remove lines that do not apply to you -->

- [X] tests and code linting passes
- [X] a test and/or benchmark is included
- [X] the commit message follows commit guidelines

##### Affected core subsystem(s)

test, cluster

##### Description of change

<!-- provide a description of the change below this comment -->

I've observed that _test-cluster-master-kill_ fails intermittently on an AIX 6.1 machine (oslevel 6100-07-08-1339) due to timeout before worker termination. There was a previous PR (https://github.com/nodejs/node-v0.x-archive/pull/9431) which arbitrarily increased the timeout for AIX to 1 second, however this value is still a guess and appears to be insufficient. It's also worth noting the arbitrary timeouts have also caused problems for other platforms, see https://github.com/nodejs/node/pull/2891#issuecomment-179158473. Arbitrary timeouts cannot compensate for external factors such as system load.

In this PR I propose removing the timeout mechanism entirely, here is how the test currently works:

1. Fork a master process from the test harness.
2. The master uses cluster.fork to create a worker.
3. The worker starts an HTTP server.
4. The master will kill itself with process.exit.
5. The worker will then also call process.exit as part of its [IPC pipe disconnect handler](https://github.com/nodejs/node/blob/8ebec086a50712bce675ae59c10ac30a4f01fa8d/lib/cluster.js#L547-L553). (*)
6. The test harness will wait an arbitrary amount of time after the master's exit to check for the child's liveness, if it is still alive the test fails.

(*) Without this mechanism the worker would become an orphan child of init.

Step 6 is inherently flaky. The test was originally added as part of https://github.com/nodejs/node-v0.x-archive/commit/94d337eb0fbf78bda2af6938b4a65d3c94808caf#diff-0faa53fc02580d5de2ebb484c41d691cR498 where it specifically tested the new disconnect pathway.
 
The test boils down to the following actions:

1. Cause the worker's disconnect handler to be run by calling exit(0) in the cluster master.
2. Insure that the disconnect handler also exits the worker process.

Since step 2 does not actually require step 1, I propose the following alternate flow:

1. Fork a master process from the test harness.
2. The master uses cluster.fork to create a worker.
3. The worker starts an HTTP server.
4. The master kills the IPC pipe (note this not the same as using the graceful cluster shutdown API in https://github.com/nodejs/node-v0.x-archive/pull/2740/commits/6c383c61611fb78d0961a332c1ad55d1d5514ba6) which causes the child to disconnect itself. Ungracefully closing the IPC pipe causes the child to execute the same disconnect handler it executes when the master calls exit(0).
5. The master waits for the child's exit event.
6. The test harness checks for the child's liveness after the master's exit event.

With this setup there is no need for the arbitrary wait time. The obvious problem is if the child never exits the test will hang - however in that case the test will still be killed by the [test harness's global timeout](https://github.com/nodejs/node/blob/8ebec086a50712bce675ae59c10ac30a4f01fa8d/tools/test.py#L587-L591).

I do believe a timeout mechanism is useful for detecting liveness issues, but the presence of arbitrary timeouts in the tests themselves should be minimized, the single timeout in the test harness suffices.

Any comments are appreciated, especially from @AndreasMadsen.

stefanmb commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/2466/

sam-github commented on test/parallel/test-cluster-master-kill.js:
```diff
@@ -21,10 +21,17 @@ if (cluster.isWorker) {
+/* Cluster.disconnect() will exercise a different 'graceful' shutdown path.
+   From the perspective of the worker, closing the channel is equivalent
+   to the parent calling process.exit(0). */
+      worker.process._channel.close();
```
I don't object, but this does use undocumented APIs, and the last test didn't. That can be annoying when trying to rearrange things that are supposed to be internal.

Why not just change the check at the end to not do a single check after some randomly chosen time interval, but to run the check every half-second, indefinitely, until it passes or the runner kills this?

stefanmb replied:
@sam-github That would be another way to do this - I'm not a huge fan of polling but it may be the lesser evil. I can do another version with your suggestion. Thanks.

stefanmb commented:
@sam-github I've adjusted the test as you suggested to simply loop indefinitely.

CI: https://ci.nodejs.org/job/node-test-pull-request/2473/

santigimeno commented:
Maybe it makes sense applying the same changes to `test-cluster-master-error`.
Other than that LGTM.

stefanmb commented:
@santigimeno Thanks for pointing that out - I've updated ```test-cluster-master-error``` with the same approach (and also adjusted some variable naming and comments in it for easier understanding). 

CI: https://ci.nodejs.org/job/node-test-pull-request/2479/

mhdawson commented:
LGTM.

santigimeno commented:
Duplicate of https://github.com/nodejs/node/pull/5056?

ncopa commented on test/parallel/test-cluster-master-kill.js:
```diff
@@ -66,15 +59,17 @@ if (cluster.isWorker) {
+    var pollWorker = function() {
+      alive = isAlive(pid);
+      if (alive) {
+        setTimeout(pollWorker, 500);
```
i think you safely can set the timeout value lower, so we don´t slow down the test up to a half second.

stefanmb commented:
@ncopa Updated with 50ms polling interval instead of 500ms. Thanks.

AndreasMadsen commented:
Nice analysis. I too is much more comfortable about polling than _just_ closing the IPC. I remember there have been issues in the past where the disconnect event did not fire when the parent died. But it's a long time ago, I could be wrong.

LGTM

stefanmb commented:
One more CI run with the last change: https://ci.nodejs.org/job/node-test-pull-request/2496/

@ncopa After the CI run I'll merge this PR and close https://github.com/nodejs/node/pull/5056 as they are largely equivalent.

stefanmb commented:
Landed as https://github.com/nodejs/node/commit/fc66e55396f9584cda2fde255bf9204926bd2cf2.

drewfish commented:
I wonder if this will fix #6193 as well.

TheAlphaNerd commented:
@stefanmb lts?

stefanmb commented:
@TheAlphaNerd I think this is OK for LTS. Thanks for reminding me.
